File: app/models/answer.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, Literal, List

from pydantic import BaseModel, Field, ConfigDict

from app.models.fetch_task import TaskRead


class AnswerGroupBase(BaseModel):
    question_id: int
    slug: Optional[str] = None
    title: str
    descriptor: Optional[str] = None
    dialogue_profile: dict = Field(default_factory=dict)


class AnswerGroupCreate(AnswerGroupBase):
    pass


class AnswerGroupRead(AnswerGroupBase):
    id: int
    created_at: datetime
    answers: List['AnswerRead'] = Field(default_factory=list)
    model_config = ConfigDict(from_attributes=True)


class AnswerBase(BaseModel):
    answer_group_id: int
    version_index: int = 1
    status: str = Field(default="draft")
    title: str
    text: str


class AnswerCreate(AnswerBase):
    pass


class AnswerRead(AnswerBase):
    id: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class AnswerUpdate(BaseModel):
    title: Optional[str] = None
    text: Optional[str] = None
    status: Optional[str] = None


class SessionBase(BaseModel):
    question_id: int
    answer_id: Optional[int] = None
    session_type: Literal["first", "review"] = "first"
    status: str = Field(default="draft")
    user_answer_draft: Optional[str] = None
    progress_state: dict = Field(default_factory=dict)


class SessionCreate(SessionBase):
    pass


class SessionRead(SessionBase):
    id: int
    started_at: datetime
    completed_at: Optional[datetime] = None


class SessionUpdate(BaseModel):
    answer_id: Optional[int] = None
    status: Optional[str] = None
    user_answer_draft: Optional[str] = None
    progress_state: Optional[dict] = None


class SessionFinalizePayload(BaseModel):
    answer_group_id: Optional[int] = None
    group_title: Optional[str] = None
    group_descriptor: Optional[str] = None
    dialogue_profile: Optional[dict] = None
    answer_title: str
    answer_text: str


class LLMConversationRead(BaseModel):
    id: int
    session_id: Optional[int] = None
    task_id: Optional[int] = None
    purpose: str
    messages: dict
    result: dict
    model_name: Optional[str] = None
    latency_ms: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnswerHistoryRead(BaseModel):
    answer: AnswerRead
    group: AnswerGroupRead
    sessions: List[SessionRead] = Field(default_factory=list)
    tasks: List[TaskRead] = Field(default_factory=list)
    conversations: List[LLMConversationRead] = Field(default_factory=list)
    review_notes_history: List[dict] = Field(default_factory=list)


class SessionHistoryRead(BaseModel):
    session: SessionRead
    tasks: List[TaskRead] = Field(default_factory=list)
    conversations: List[LLMConversationRead] = Field(default_factory=list)

File: frontend/src/types/answer.ts
import type { Session } from './session';
import type { FetchTask } from './question';

export interface Answer {
  id: number;
  answer_group_id: number;
  version_index: number;
  status: string;
  title: string;
  text: string;
  created_at: string;
}

export interface AnswerGroup {
  id: number;
  question_id: number;
  slug?: string | null;
  title: string;
  descriptor?: string | null;
  direction_descriptor?: string | null;
  dialogue_profile?: Record<string, unknown>;
  created_at: string;
  answers: Answer[];
}

export interface Sentence {
  id: number;
  paragraph_id: number;
  order_index: number;
  text: string;
  translation_en?: string | null;
  translation_zh?: string | null;
  difficulty?: string | null;
  extra: Record<string, unknown>;
  created_at: string;
  chunks: SentenceChunk[];
}

export interface Lexeme {
  id: number;
  headword: string;
  sense_label?: string | null;
  gloss?: string | null;
  translation_en?: string | null;
  translation_zh?: string | null;
  pos_tags?: string | null;
  difficulty?: string | null;
  hash: string;
  extra: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface ChunkLexemeUsage {
  id: number;
  chunk_id: number;
  lexeme_id: number;
  order_index: number;
  role?: string | null;
  lexeme: Lexeme;
}

export interface SentenceChunk {
  id: number;
  sentence_id: number;
  order_index: number;
  text: string;
  translation_en?: string | null;
  translation_zh?: string | null;
  chunk_type?: string | null;
  extra?: Record<string, unknown>;
  lexemes: ChunkLexemeUsage[];
}

export interface Paragraph {
  id: number;
  answer_id: number;
  order_index: number;
  role_label?: string | null;
  summary?: string | null;
  extra: Record<string, unknown>;
  created_at: string;
  sentences: Sentence[];
}

export interface LLMConversationLog {
  id: number;
  session_id: number | null;
  task_id: number | null;
  purpose: string;
  messages: Record<string, unknown>;
  result: Record<string, unknown>;
  model_name?: string | null;
  latency_ms?: number | null;
  created_at: string;
}

export interface AnswerHistory {
  answer: Answer;
  group: AnswerGroup;
  sessions: Session[];
  tasks: FetchTask[];
  conversations: LLMConversationLog[];
}
